backend/headless/backend.c

```diff
@@ -220,21 +225,24 @@ struct wlr_backend *wlr_headless_backend_create(struct wl_display *display) {
 		goto error_dup;
 	}
 
+#if WLR_HAS_GBM_SUPPORT
 	struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
 	if (gbm_alloc == NULL) {
 		wlr_log(WLR_ERROR, "Failed to create GBM allocator");
 		close(drm_fd);
 		goto error_dup;
 	}
+	base_allocator = &gbm_alloc->base;
+#endif
```

Comment:
I don't normally mind `#ifdef`'s, but I feel like there's a lot of repetition throughout this change.

Options:
- A private, shared helper (`wlr_gbm_allocator_try_create`?) where the `#ifdefs` go.
- Something like a `wlr_allocator_autocreate`
- Let `wlr_gbm_allocator_create` exist but fail when GBM is unavailable - this approach could also work for other components, such as libinput and gles.

@emersion what do you think?

Reply:
> Let wlr_gbm_allocator_create exist but fail when GBM is unavailable - this approach could also work for other components, such as libinput and gles.

This doesn't let us make a difference between "dependency is missing" and "something went wrong".

> Something like a wlr_allocator_autocreate

This seems like a good idea. There are interactions with https://github.com/swaywm/wlroots/pull/2737#discussion_r602707670 here: the `wlr_allocator_autocreate` caller doesn't know anymore whether the allocator will use shm or DMA-BUFs. The caller also can't guarantee the allocator and the renderer will work fine together. If we keep the new function private for now, we could sort out these issues later.

Reply:
> This doesn't let us make a difference between "dependency is missing" and "something went wrong".

That's true. I imagine the states would only be unavailable, broken dependency and allocation error. If proceeding to the next available allocator is acceptable, then we could just log the cause from the _create method.

> [regarding wlr_allocator_autocreate] If we keep the new function private for now, we could sort out these issues later.

I'd ACK that approach. @hizukiayaka, does that work for you?

Reply:
`wlr_allocator_autocreate` would be used as we all agreed, neither gbm allocator nor drm dumb header files would be exported anymore.
Once we don't the part how to deal with the buffer its allocate or even the part the buffer it imported, we could remove all the `#ifdef` here